src/types/widget.ts

```diff
@@ -0,0 +1,23 @@
+import {
+  Helper,
+  HelperState,
+  SearchResults,
+  InstantSearch,
+  SearchParameters,
+} from './instantsearch';
+
+interface WidgetArgs {
+  helper: Helper;
+  state?: HelperState;
+  results?: SearchResults;
+  instantSearchInstance: InstantSearch;
+}
+
+export interface Widget {
+  init(options: WidgetArgs): void;
```

Comment:
This is not entirely true a widget can have only a `render` or an `init` method.

https://github.com/algolia/instantsearch.js/blob/3de59a32a08df71e0d4ba25af060560d4d6495ec/src/lib/InstantSearch.js#L144

Reply:
A widget can also have both, but at least one of them.

How would you translate that to TypeScript?

Reply:
A discriminate union between two types with it required `type Widget = ({ init(): void } | { render(): void }) & { ...commonStuff }`